third_party/move/move-compiler/src/expansion/translate.rs

```diff
@@ -10,8 +10,8 @@ use crate::{
     expansion::{
         aliases::{AliasMap, AliasSet},
         ast::{
-            self as E, Address, Fields, LValueOrDotDot_, ModuleAccess_, ModuleIdent, ModuleIdent_,
-            SpecId,
+            self as E, Address, Fields, LValueOrDotDot_, LValue_, ModuleAccess_, ModuleIdent,
```

Comment:
What is the rationale for doing this desugaring in this expansion phase?

The index notation and receiver-style calls are desugared in the move-model generation phase. The for loop expansion was done in the parser as a stop-gap to make it also work for compiler v1.

Reply:
The index notation and receiver calls need type information for automatic (de)referencing, so have to wait until the model generation phase. We don't need type information here. Desugaring in the model generation phase also requires us to adapt the type-checking rules, making the transformation harder. I pasted the `exp_builder` code which we would need to change
```rust
EA::Exp_::Assign(lhs, rhs) => {
                self.require_impl_language(&loc);
                let (rhs_ty, rhs) = self.translate_exp_free(rhs);
                let lhs = self.translate_lvalue_list(
                    lhs,
                    &rhs_ty,
                    WideningOrder::RightToLeft,
                    true, /*match_locals*/
                    &ErrorMessageContext::Assignment,
                );
                self.check_duplicate_assign(&lhs);
                // The type of the assign is Unit
                let result_ty = self.check_type(&loc, &Type::unit(), expected_type, context);
                let id = self.new_node_id_with_type_loc(&result_ty, &loc);
                let lhs_ty = self.env().get_node_type(lhs.node_id());
                let rhs_ty = self
                    .subs
                    .specialize(&self.env().get_node_type(rhs.node_id()));
                let rhs = rhs.into_exp();
                // Insert freeze for rhs of the assignment
                let rhs = if lhs_ty.is_tuple()
                    && rhs_ty.is_tuple()
                    && matches!(rhs.as_ref(), ExpData::Call(_, Operation::Tuple, _))
                {
                    if let (Pattern::Tuple(_, lhs_pats), Type::Tuple(rhs_tys)) = (&lhs, &rhs_ty) {
                        let lhs_tys = lhs_pats
                            .iter()
                            .map(|pat| self.get_node_type(pat.node_id()))
                            .collect_vec();
                        self.freeze_tuple_exp(&lhs_tys, rhs_tys, rhs, &loc)
                    } else {
                        self.try_freeze(&lhs_ty, &rhs_ty, rhs)
                    }
                } else {
                    self.try_freeze(&lhs_ty, &rhs_ty, rhs)
                };
                ExpData::Assign(id, lhs, rhs)
            },
            EA::Exp_::Mutate(lhs, rhs) => {
                let (rhs_ty, rhs) = self.translate_exp_free(rhs);
                // Do not freeze when translating the lhs of a mutate operation
                self.insert_freeze = false;
                let (lhs_ty, lhs) = if let EA::Exp_::Index(target, index) = &lhs.value {
                    let result_ty =
                        Type::Reference(ReferenceKind::Mutable, Box::new(rhs_ty.clone()));
                    if let Some(call) = self.try_resource_or_vector_index(
                        &loc, target, index, context, &result_ty, true,
                    ) {
                        (result_ty, call)
                    } else {
                        self.translate_exp_free(lhs)
                    }
                } else {
                    self.translate_exp_free(lhs)
                };
                self.insert_freeze = true;
                self.check_type(
                    &self.get_node_loc(lhs.node_id()),
                    &Type::Reference(ReferenceKind::Mutable, Box::new(rhs_ty)),
                    &lhs_ty,
                    &ErrorMessageContext::Assignment,
                );
                let result_ty = self.check_type(&loc, &Type::unit(), expected_type, context);
                let id = self.new_node_id_with_type_loc(&result_ty, &loc);
                ExpData::Mutate(id, lhs.into_exp(), rhs.into_exp())
            },
```
This is already complicated.

We can't do it in the parser either for two reasons.
1. We can't generate non-user symbols, since the validness of variable names is checked in the expansion phase.
2. We need to check the LHS of an assignment for the transformation, otherwise, we may have `1 += 1` transformed into `let t = &mut; *t += 1;` which becomes valid code... And whether the LHS of an assignment is a proper lvalue is done in the expansion phase.